compiler/noirc_frontend/src/elaborator/mod.rs

```diff
@@ -417,12 +406,42 @@ impl<'context> Elaborator<'context> {
         meta.function_body = FunctionBody::Resolved;
 
         self.trait_bounds.clear();
-        self.type_variables.clear();
         self.interner.update_fn(id, hir_func);
         self.current_function = old_function;
         self.current_item = old_item;
     }
 
+    /// Defaults all type variables used in this function context then solves
+    /// all still-unsolved trait constraints in this context.
+    fn check_and_pop_function_context(&mut self) {
+        if let Some(context) = self.function_context.pop() {
```

Comment:
Should this panic if `function_context` is empty? (it would mean we forgot to push to it)

Reply:
I was trying to avoid panics but I suppose we do want to panic in that case - I'll change it

Reply:
Fixed